### PR TITLE
py: Fix explicit allocation for empty namedtuple

### PR DESCRIPTION
**Pull request by LarsH**
The empty tuple is usually a constant object, but named tuples
must be allocated to allow modification.

Also added a regression test to verify creating an empty named tuple works.

Fixes issue #7870.

### CODE REVIEW

File: py/objnamedtuple.c [hunk 1]
```diff
@@ -109,7 +109,14 @@ STATIC mp_obj_t namedtuple_make_new(const mp_obj_type_t *type_in, size_t n_args,
     }
 
     // Create a tuple and set the type to this namedtuple
-    mp_obj_tuple_t *tuple = MP_OBJ_TO_PTR(mp_obj_new_tuple(num_fields, NULL));
+    mp_obj_tuple_t *tuple;
+    if (num_fields == 0) {
+        // The empty tuple is a constant object; explicit allocation needed
+        tuple = m_new_obj_var(mp_obj_tuple_t, mp_obj_t, 0);
```
Comment (dlech) [inline py/objnamedtuple.c:115]: Does `m_new_obj_far` work for the non-empty case too? If yes, we can avoid the if statement.

Comment (LarsH) [reply py/objnamedtuple.c:115]: It should work for the non-empty case too, as it is what `mp_obj_new_tuple` calls internally.

This is my first contribution; what is the best way to act? Should I force-push that change to my branch, create a new branch or create a new commit?

Comment (dlech) [reply py/objnamedtuple.c:115]: Since this is a small change, I would just force-push to this branch.

File: py/objnamedtuple.c [hunk 2]
```diff
@@ -109,7 +109,9 @@ STATIC mp_obj_t namedtuple_make_new(const mp_obj_type_t *type_in, size_t n_args,
     }
 
     // Create a tuple and set the type to this namedtuple
-    mp_obj_tuple_t *tuple = MP_OBJ_TO_PTR(mp_obj_new_tuple(num_fields, NULL));
+    // Explicit allocation; mp_obj_new_tuple with num_fields=0 returns read-only object
+    mp_obj_tuple_t *tuple = mp_obj_malloc_var(mp_obj_tuple_t, mp_obj_t, num_fields, &mp_type_tuple);
+    tuple->len = num_fields;
     tuple->base.type = type_in;
```
Comment (dlech) [inline py/objnamedtuple.c:115]: I think we could pass `type_in` to `mp_obj_malloc_var()` directly and delete this line to override the type.

Comment (dpgeorge) [reply py/objnamedtuple.c:115]: Yes that would be good!

Comment (LarsH) [reply py/objnamedtuple.c]: Great suggestion, changed that in d2da4b4

**Comment by dpgeorge** [comment]
Thanks for updating.  Rebased and merged in 5bf3765631e645412dbd62a616cfdadeca5ea0c3
